### PR TITLE
feature: added pure C API for setting upstream request timeouts

### PR DESCRIPTION
**Pull request by agentzh**
This supersedes PR #700.

### CODE REVIEW

**Comment by agentzh** [comment]
This PR can optionally take advantage of this nginx core patch: https://github.com/openresty/openresty/blob/master/patches/nginx-1.9.15-upstream_timeout_fields.patch

Still we can work with the stock nginx core, just requiring a bit more memory per request.

**Comment by agentzh** [comment]
@xbsura @toruneko Please review this PR. Thanks!

File: src/ngx_http_lua_balancer.c [hunk 1]
```diff
@@ -537,6 +541,101 @@ ngx_http_lua_ffi_balancer_set_current_peer(ngx_http_request_t *r,
...
+            return NGX_ERROR;
+        }
+
+        ngx_memcpy(ucf, u->conf, sizeof(ngx_http_upstream_conf_t));
+
+        u->conf = ucf;
+        bp->cloned_upstream_conf = 1;
+
+    } else {
+        ucf = u->conf;
```
Comment (toruneko) [inline src/ngx_http_lua_balancer.c]: hah good job, it means we can invoke this function twice，but  only alloc memory one times。

**Comment by toruneko** [comment]
I review this PR, and not find any problem. @xbsura ？

File: src/ngx_http_lua_balancer.c (hunk 1 above)
Comment (agentzh) [inline src/ngx_http_lua_balancer.c]: @toruneko Yup. That's right :)

Comment (guanglinlv) [inline src/ngx_http_lua_balancer.c]: @agentzh , someone maybe want to set different timeout for different peer when retrying the next one. how about  keep it left to the user?

**Comment by xbsura** [comment]
oh, good
meeting pre work, not in work, sorry

**Comment by toruneko** [comment]
I find a problem that if we do not use this api to set timeouts, and the u->*_timeout will be blank if patched    nginx-1.9.15-upstream_timeout_fields.patch.
we should patches ngx_http_proxy_handler function after 
`u->conf = &plcf->upstream; ` like
```
u->conf = &plcf->upstream;
u->read_timeout = u->conf->read_timeout;
u->send_timeout = u->conf->send_timeout;
u->connect_timeout = u->conf->connect_timeout;
```

**Comment by agentzh** [comment]
@toruneko It's already covered in OpenResty's nginx core patch here:

https://github.com/openresty/openresty/blob/master/patches/nginx-1.9.15-upstream_timeout_fields.patch#L17

BTW, patching ngx_proxy for this is wrong since there are unlimited number of other upstream modules out there, including standard upstream modules like ngx_fastcgi and ngx_uwsgi, not to mention 3rd-party upstream modules like ngx_memc and ngx_redis2.

**Comment by agentzh** [comment]
This PR is already merged into master. If you find further problems, please create separate github issues or PRs. Thank you.

**Pull request closed without merging**